demos/src/ann_benchmark.rs

```diff
@@ -74,165 +103,159 @@ const DATASETS: [(&str, &str, DistanceType); 11] = [
 ];
 
 fn main() {
+    let args = Args::parse();
+
     for dataset in &DATASETS {
+        if !dataset.0.contains(&args.dataset) {
+            continue;
+        }
+
         let mut data = AnnBenchmarkData::new(dataset.0, dataset.1);
+        println!("Original data size: {}", data.vectors_count * data.dim * 4);
         if dataset.2 == DistanceType::Dot {
             data.cosine_preprocess();
         }
 
-        let encoded = data.encode_data(dataset.2, None);
-
-        let permutor = permutation_iterator::Permutor::new(data.vectors_count as u64);
-        let random_indices: Vec<u32> = permutor.map(|i| i as u32).collect();
-        let linear_indices: Vec<u32> = (0..data.vectors_count as u32).collect();
-
-        #[cfg(target_arch = "x86_64")]
-        {
-            println!("Measure AVX2 linear access");
-            data.measure_scoring(data.queries_count / 10, |query, scores| {
-                for &index in &linear_indices {
-                    scores[index as usize] = unsafe {
-                        dot_avx(query, data.vectors.row(index as usize).as_slice().unwrap())
-                    };
-                }
-            });
-        }
+        let timer = std::time::Instant::now();
+        let data_iter = data
+            .vectors
+            .rows()
+            .into_iter()
+            .map(|row| row.to_slice().unwrap());
+        let vector_parameters = VectorParameters {
+            dim: data.dim,
+            count: data.vectors_count,
+            distance_type: dataset.2,
+            invert: false,
+        };
 
-        #[cfg(target_arch = "x86_64")]
-        {
-            println!("Measure Quantized AVX2 linear access");
-            data.measure_scoring(data.queries_count / 10, |query, scores| {
-                let query_u8 = encoded.encode_query(query);
-                for &index in &linear_indices {
-                    scores[index as usize] = encoded.score_point_avx(&query_u8, index);
-                }
-            });
-        }
+        match args.method.as_str() {
+            "pq" => {
```

Comment:
Since these two branches are _almost_ identical, we could box `TEncodedVectors`. Or did you do this on purpose to generate `run_test` for each type?

Incomplete suggestion with boxing:

```rust
println!("Start encoding:");
match args.method.as_str() {
    "pq" => Box::new(
        EncodedVectorsPQ::encode(
            data_iter,
            Vec::<u8>::new(),
            &vector_parameters,
            args.chunk_size,
            num_cpus::get(),
        )
        .unwrap(),
    ),
    "u8" => Box::new(
        EncodedVectorsU8::encode(
            data_iter,
            Vec::<u8>::new(),
            &vector_parameters,
            args.quantile,
        )
        .unwrap(),
    ),
    _ => panic!("Unknown method"),
};
println!("encoding time: {:?}", timer.elapsed());
run_test(&data, &encoded, dataset.2, &args);
```

Reply:
This file about demo of quantization. By arguments, user can test quantization on different datasets from http://ann-benchmarks.com. It measures KNN accuracy and benches KNN performance. Because it's benchmark too, I don't want to box encoded storage because I don't want to pay dynamic call for each distance measurement

Reply:
Makes sense. Thanks for explaining!